langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiStreamingChatModel.java

```diff
@@ -152,22 +158,52 @@ private void generate(List<ChatMessage> messages,
 
         ChatCompletionRequest request = requestBuilder.build();
 
+        ChatLanguageModelRequest observabilityRequest = createObservabilityRequest(request, messages, toolSpecifications);
+        if (listener != null) {
+            listener.onRequest(observabilityRequest);
+        }
+
         int inputTokenCount = countInputTokens(messages, toolSpecifications, toolThatMustBeExecuted);
         OpenAiStreamingResponseBuilder responseBuilder = new OpenAiStreamingResponseBuilder(inputTokenCount);
 
+        AtomicReference<String> responseId = new AtomicReference<>();
+        AtomicReference<String> responseModel = new AtomicReference<>();
+
         client.chatCompletion(request)
                 .onPartialResponse(partialResponse -> {
                     responseBuilder.append(partialResponse);
                     handle(partialResponse, handler);
+
+                    if (!isNullOrBlank(partialResponse.id())) {
+                        responseId.set(partialResponse.id());
+                    }
+                    if (!isNullOrBlank(partialResponse.model())) {
+                        responseModel.set(partialResponse.model());
+                    }
                 })
                 .onComplete(() -> {
                     Response<AiMessage> response = responseBuilder.build(tokenizer, toolThatMustBeExecuted != null);
                     if (!isOpenAiModel) {
                         response = removeTokenUsage(response);
                     }
+
+                    if (listener != null) {
+                        ChatLanguageModelResponse observabilityResponse = createObservabilityResponse(
+                                responseId.get(),
+                                responseModel.get(),
+                                response
+                        );
+                        listener.onResponse(observabilityRequest, observabilityResponse);
+                    }
+
                     handler.onComplete(response);
                 })
-                .onError(handler::onError)
+                .onError(error -> {
+                    if (listener != null) {
+                        listener.onError(observabilityRequest, error);
```

Comment:
TODO:
- check that exception cannot be thrown outside of onError
- what to do with exceptions produced by user's handler?

Reply:
Swallow them and log a warning.
Monitoring code shouldn't break the main business logic flow.

Reply:
> Monitoring code shouldn't break the main business logic flow.

Very true, but I see this as a generic middleware piece and theoretically some other kind of listener might want to do something with it